include/util_quda.h

```diff
@@ -7,15 +7,42 @@
 #include <tune_key.h>
 #include <malloc_quda.h>
 
+#ifndef QUDA_MUSTTAIL
+#ifdef __clang__
+#define QUDA_MUSTTAIL __attribute__((musttail))
+#else
+#define QUDA_MUSTTAIL
+#endif
+#endif
+
 namespace quda
 {
   // strip path from __FILE__
   constexpr const char *str_end(const char *str) { return *str ? str_end(str + 1) : str; }
   constexpr bool str_slant(const char *str) { return *str == '/' ? true : (*str ? str_slant(str + 1) : false); }
   constexpr const char *r_slant(const char *str) { return *str == '/' ? (str + 1) : r_slant(str - 1); }
   constexpr const char *file_name(const char *str) { return str_slant(str) ? r_slant(str_end(str)) : str; }
+
+  template<int A, int B, int D=1, typename F>
```

Comment:
Probably better to rename these variables to make them more intuitive (`lower` instead of `A`, `upper` instead of `B`) and add comments on how this works.

Reply:
Yes.  Once we understand the full impact on the code gen.